File: cpp/problem404.cpp

// Class: Solution, iteratively
// Creat: Yangyao Chen, 2018/10/15
class Solution {
public:
    int sumOfLeftLeaves(TreeNode* root) {
        int sum = 0;
        stack<TreeNode *> stk;
        while( root || !stk.empty() ){
            while(root) {
                if( root->left && ! root->left->left && ! root->left->right ) sum += root->left->val;
                if( root->right ) stk.push( root->right );
                root = root->left;
            }
            if( !stk.empty() ){
                root = stk.top(); stk.pop();
            }
        }
        return sum;
    }
};


// Class: Solution, recursively
// Creat: Yangyao Chen, 2018/10/15
class Solution {
public:
    int sumOfLeftLeaves(TreeNode* root) {
        if( !root ){ return 0; }
        return sumOfLeftLeaves(root, false);
    }
    int sumOfLeftLeaves(TreeNode* root, bool isleft) {
        if( isleft && !root->left && !root->right ) return root->val;
        else{
            int sum = 0;
            if( root->left ) sum += sumOfLeftLeaves(root->left, true);
            if( root->right ) sum += sumOfLeftLeaves(root->right, false);
            return sum;
        }
    }
};

File: swift/problem404.swift
public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }
}

class Solution {
    func sumOfLeftLeaves(_ root: TreeNode?) -> Int {
        var tmp = root
        var stack = [TreeNode]()
        var sum = 0
        var flag = false
        
        while (stack.count > 0 || tmp != nil) {
            flag = false
            if (tmp?.left != nil) {
                flag = true
            }
            while (tmp != nil) {
                stack.append(tmp!)
                tmp = tmp!.left
            }
            tmp = stack.removeLast()
            if (flag && tmp!.right == nil) {
                sum += tmp!.val
            }
            tmp = tmp!.right
        }
        
        return sum
    }
}
